src/tools/pub/pubspec.md

```diff
@@ -269,6 +273,65 @@ For more information, see
 [pub global](/tools/pub/cmd/pub-global#running-a-script-from-your-path).
 
 
+### Platforms
+
+When publishing packages to the [pub.dev site.]({{site.pub}}) the set of
+_supported platforms_ will be automatically detected based on transitively
+imported `dart:...` libraries and [Flutter plugin declarations][]. The set of
+supported platforms is displayed on the package page, and used for filtering
+when searching for packages on the [pub.dev site.]({{site.pub}}).
+
+If the supported platforms automatically detected is incorrect, the list of
+_supported platforms_ can be explicitly declared using the `platforms:` key
+as follows:
+
+{% prettify yaml tag=pre+code %}
+# This package supports all platforms listed below
+platforms:
+  android:
+  ios:
+  linux:
+  macos:
+  web:
+  windows:
+{% endprettify %}
+
+The automatic detection of supported platforms is not perfect. For example, a
+package using `chmod` through `Process.run` from `dart:io` won't work on
+Windows, even though `dart:io` is available on Windows. Hence, it is reasonable
+to explicitly declare that the package only supports Linux and MacOS, as
+illustrated below:
+
+{% prettify yaml tag=pre+code %}
+# This package only supports Linux and MacOS
+platforms:
+  linux:
+  macos:
+{% endprettify %}
+
+{{site.alert.info}}
+  The automatic detection of support platforms will conclude that a package does
+  not support web, if the primary library transitively imports `dart:io`.
+  Because while it possible to import `dart:io` when compiling Dart to
+  Javascript, it is not possible to use functionality from `dart:io` at runtime
+  when compiled to Javascript. Trying to use `dart:io` when compiled to
+  Javascript will throw an error.
+
+  Some packages do import `dart:io` and avoid using it when they are compiled to
+  Javascript. This technically works, and but prevents automatic platform
+  detection from correctly determining that the package supports web.
+
+  These packages are encouraged to [conditionally import][] `dart:io` instead of
+  explicitly declaring supported platforms. Using conditional import reduces the
+  risk of runtime errors due to accidental usage of `dart:io` when compiled to
+  Javascript, and it ensures that packages with dependency upon this package
+  doesn't also have to explicitly declare supported platforms.
+{{site.alert.end}}
```

Comment:
This note is a bit long... I'm not sure if we can phrase it shorter... if we should omit it?

Or if we should wrap it in a [`<details><summary>`](https://gist.github.com/scmx/eca72d44afee0113ceb0349dd54a84a2) to make it expandable.

The goal is really just to explain one of the common pitfalls, because `dart:io` is weird... compared to `dart:html` which cannot be imported when running on the VM.

Reply:
I could also move this to a page dedicated to "platform detection". Just because we have such a page doesn't mean we have to include in the right hand menu... we could just reference from here, and leave it for people to find it that way or through google :D

Reply:
I think I like the idea of moving it to a separate page as this seems like a pretty advanced piece of functionality and we can link to it from other locations discussing conditional imports. @mit-mit What do you think?